src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlBinaryWriter.cs

```diff
@@ -1437,7 +1442,7 @@ public override unsafe void WriteArray(string? prefix, string localName, string?
 
         public override unsafe void WriteArray(string? prefix, XmlDictionaryString localName, XmlDictionaryString? namespaceUri, decimal[] array, int offset, int count)
         {
-            if (Signing)
```

Comment:
Based on the code it is very difficult to known if it produces the same result or if the format for arrays have changed.
You should ensure there are tests to cover this similar to those is https://github.com/dotnet/runtime/pull/71478 , but it might make sense to check more types than I did. 

My concern is that currently it only check round-trip support and I think it is important to ensure that the produced file will be readable on LittleEndian and that it can read content produced by little endian machins.

Suggestion: 
If more code is needed then maybe you want to consider writing a genereic WriteArray helper.

My own idea (which assumed that it would have been ok to update the writer PR or that it was merged first)  to solve the problem was to  
* use a genereic ReadArray/WriteArray helper with a loop for reading/writing elements on big endian. 
* generic "ReverseEndianness" method which need to handle at least short, int, long (other primitives,  except deicmal, can always be cast to these using Unsafe or MemoryMarshal.Cast) 
   -  (Read/WriteRawBytes might have been able to do raw byte swapping as well)

Reply:
Thanks for the thorough review!   You're right that the array cases were not correct.  I though we could just fall back to the base class implementation like in the reader, but in the writer the base class actually doesn't create any array records, just a series of "regular" records.   This is probably not _wrong_, but it clearly would be preferable to create the same output on big-endian as on little-endian.

Normally, I agree we should just merge your rewrite first, and then fix the endian problems on top.  However, I'm not sure the rewrite will still be accepted into net7 at this stage, and I'd really like to see the endian bugs fixed in net7, so I'd prefer to have the endian fix go first, and the performance rewrite on top of that.

The updated implementation I just pushed replaced the generic `UnsafeWriteArray` helper with per-type array helpers along the line of the existing `WriteDateTimeArray` etc.  These new helpers are still unsafe (for now) since I left the little-endian implementation unchanged, but with your patch on top it should be straightforward to make them all safe.

As to the `ReverseEndianness` suggestion, I feel it should be cleaner to eliminate explicit byte-swap operations in favor of type-specific fixed-byte-order operations from `BinaryPrimitives` - see also the comment by @jkotas above.